PS-9139 [DOCS] - Document new PS Encryption UDFs functionality - 8.0 (#419)
percona/psmysql-docs: docs/encryption-functions.md

# Encryption user-defined functions

The encryption user-defined functions (UDF) let you encrypt and decrypt data. You can choose different encryption algorithms and manage the range of data to encrypt.

## Version updates

> Percona Server for MySQL 8.0.40 adds the following:

Reviewer: ```suggestion
Percona Server for MySQL 8.0.41 adds the following:
```

* Support for `PKCS1 OAEP` padding for RSA encrypt and decrypt operations

    <details>

>        <summary> `PKCS1 OAEP` padding explanation</summary>

Reviewer: May be add links to Wikipedia
https://en.wikipedia.org/wiki/PKCS_1
https://en.wikipedia.org/wiki/Optimal_asymmetric_encryption_padding
https://en.wikipedia.org/wiki/Probabilistic_signature_scheme

# Encryption user-defined functions

The encryption user-defined functions (UDF) let you encrypt and decrypt data. You can choose different encryption algorithms and manage the range of data to encrypt.

## Version updates

Percona Server for MySQL 8.0.41 adds the following:

* Support for `pkcs1`, `oaep`, or `no` padding for RSA encrypt and decrypt operations

    <details>
       <summary> `pkcs1` padding explanation</summary>

>         The `pkcs1` padding prevents patterns that attackers could exploit by including a random sequence of bytes which ensures that the ciphertext is different no matter how many times it is encrypted.

Reviewer: ```suggestion
        The [`RSAES-PKCS1-v1_5`](https://en.wikipedia.org/wiki/PKCS_1) RSA encryption padding scheme prevents patterns that attackers could exploit by including a random sequence of bytes which ensures that the ciphertext is different no matter how many times it is encrypted.
```

    </details>
    
    <details>
       <summary> `oeap` padding explanation</summary>  

>         The `oeap` (Optimal Asymmetric Encryption Padding) padding adds a randomized mask generation function. This function makes it more difficult for attackers to exploit weaknesses in the encryption algorithm or to recover the original message.

Reviewer: ```suggestion
        The [`RSAES-OAEP`](https://en.wikipedia.org/wiki/PKCS_1)  - [`Optimal Asymmetric Encryption Padding`](https://en.wikipedia.org/wiki/Optimal_asymmetric_encryption_padding) RSA encryption padding scheme adds a randomized mask generation function. This function makes it more difficult for attackers to exploit weaknesses in the encryption algorithm or to recover the original message.
```

    </details>
    
    <details>
       <summary> `no` padding explanation</summary>
       Using `no` padding means that the plaintext message is encrypted without adding an extra layer before performing the RSA encryption operation.
    </details>

> * Support for `PKCS1 PSS` padding for RSA sign and verify operations

Reviewer: ```suggestion
* Support for `pkcs1` or `pkcs1_pss` padding for RSA sign and verify operations
```

    <details>
       <summary> `PKCS1 PSS` padding explanation</summary>
        PKCS PSS (Probabilistic Signature Scheme) is a cryptographic algorithm used to add randomness to a message before signing it with a private key. This randomness helps to increase the security of the signature and make it more resistant to various attacks. 

>     </details>

Reviewer: ```suggestion
    <details>
       <summary> `pkcs1` padding explanation</summary>
        The [`RSASSA-PKCS1-v1_5`](https://en.wikipedia.org/wiki/PKCS_1) is a deterministic RSA signature padding scheme that hashes a message, pads the hash with a specific structure, and encrypts it with the signer's private key for signature generation. 
    </details>

    <details>
       <summary> `pkcs1_pss` padding explanation</summary>
        The [`RSASSA-PSS`](https://en.wikipedia.org/wiki/PKCS_1) - [`Probabilistic Signature Scheme'](https://en.wikipedia.org/wiki/Optimal_asymmetric_encryption_padding) is an RSA signature padding scheme used to add randomness to a message before signing it with a private key. This randomness helps to increase the security of the signature and make it more resistant to various attacks. 
    </details>
```

The following are the function’s parameters:

> * algorithm - the encryption algorithm supports RSA to decrypt the string. Percona Server for MySQL 8.0.40 added support for PKCS1 OAEP padding for RSA encrypt operations

Reviewer: Starting from 8.0.41 these functions `asymmetric_decrypt()` / `asymmetric_encrypt()`  accept one more optional parameter `padding` when `algorithm` is `RSA`
`padding` - the padding scheme used for encryption/decryption (can be one of the `no` / `pkcs1` / `oaep`). If omitted, its value will be implicitly determined based on the value of the `encryption_udf.legacy_padding_scheme` variable.

* algorithm - the encryption algorithm supports RSA to decrypt the string.

> * padding - An optional parameter introduced in Percona Server for MySQL 8.0.41. It is used with the RSA algorithm and supports padding schemes like `no`, `pkcs1`, or `oaep`. If you skip this parameter, the system determines its value based on the [`encryption_udf.legacy_padding_scheme`](#encryption_udf.legacy_padding_scheme) variable.

Reviewer: Move `padding` after `key_str`

Reviewer: Also, `crypt_str` is not described here. (this concerns original 8.0.28 implementation).

> * padding - An optional parameter introduced in Percona Server for MySQL 8.0.41. It is used with the RSA algorithm and supports padding schemes like `no`, `pkcs1`, or `oaep`. If you skip this parameter, the system determines its value based on the [`encryption_udf.legacy_padding_scheme`](#encryption_udf.legacy_padding_scheme) variable.

Reviewer: ```suggestion
* padding - An optional parameter introduced in Percona Server for MySQL 8.0.41. It is used with the RSA algorithm and supports RSA encryption padding schemes like `no`, `pkcs1`, or `oaep`. If you skip this parameter, the system determines its value based on the [`encryption_udf.legacy_padding_scheme`](#encryption_udf.legacy_padding_scheme) variable.
```

* algorithm - the encryption algorithm supports RSA to encrypt the string.

> * padding - An optional parameter introduced in Percona Server for MySQL 8.0.41. It is used with the RSA algorithm and supports padding schemes like `no`, `pkcs1`, or `oaep`. If you skip this parameter, the system determines its value based on the [`encryption_udf.legacy_padding_scheme`](#encryption_udf.legacy_padding_scheme) variable.

Reviewer: Move `padding` after `key_str`

> * padding - An optional parameter introduced in Percona Server for MySQL 8.0.41. It is used with the RSA algorithm and supports padding schemes like `no`, `pkcs1`, or `oaep`. If you skip this parameter, the system determines its value based on the [`encryption_udf.legacy_padding_scheme`](#encryption_udf.legacy_padding_scheme) variable.

Reviewer: ```suggestion
* padding - An optional parameter introduced in Percona Server for MySQL 8.0.41. It is used with the RSA algorithm and supports RSA encryption padding schemes like `no`, `pkcs1`, or `oaep`. If you skip this parameter, the system determines its value based on the [`encryption_udf.legacy_padding_scheme`](#encryption_udf.legacy_padding_scheme) variable.
```

The parameters are the following:

> * algorithm - the encryption algorithm supports either RSA or DSA to encrypt the string. Percona Server for MySQL 8.0.40 added support for PKCS1 PSS padding for RSA sign operations.

Reviewer: Starting from 8.0.41 these functions asymmetric_sign() / asymmetric_verify() accept one more optional parameter `padding` when algorithm is `RSA`
`padding` - the padding scheme used for sign/verify (can be one of the` pkcs1` / `pkcs1_pss`). If omitted, its value will be implicitly determined based on the value of the encryption_udf.legacy_padding_scheme variable.

The parameters are the following:

> * algorithm - the encryption algorithm supports either RSA or DSA to encrypt the string. 

Reviewer: Also add
```
* padding - An optional parameter introduced in Percona Server for MySQL 8.0.41. It is used with the RSA algorithm and supports RSA signature padding schemes like `pkcs1`, or `pkcs1_pss`. If you skip this parameter, the system determines its value based on the [`encryption_udf.legacy_padding_scheme`](#encryption_udf.legacy_padding_scheme) variable.
```
as the last parameter for this function.

The parameters are the following:

> * algorithm - supports either ‘RSA’ or ‘DSA’. Percona Server for MySQL 8.0.40 added support for PKCS1 PSS padding for RSA verify operations.

Reviewer: ```suggestion
* algorithm - supports either ‘RSA’ or ‘DSA’.
```

Reviewer: Also add
```
* padding - An optional parameter introduced in Percona Server for MySQL 8.0.41. It is used with the RSA algorithm and supports RSA signature padding schemes like `pkcs1`, or `pkcs1_pss`. If you skip this parameter, the system determines its value based on the [`encryption_udf.legacy_padding_scheme`](#encryption_udf.legacy_padding_scheme) variable.
```
as the last parameter for this function.

| Option       | Description      |
|--------------|------------------|
| command-line | Yes              |
| scope        | Global           |
| data type    | Boolean |
| default      | OFF            |

This system variable is a BOOLEAN type and is set to `OFF` by default. 

This variable controls how the functions `asymmetric_encrypt()`, `asymmetric_decrypt()`, `asymmetric_sign()`, and `asymmetric_verify()` behave when you don’t explicitly set the padding parameter.

•	When encryption_udf.legacy_padding_scheme is OFF:

    • asymmetric_encrypt() and asymmetric_decrypt() use OAEP padding.

    • asymmetric_sign() and asymmetric_verify() use PKCS1_PSS padding.

•	When encryption_udf.legacy_padding_scheme is ON:

    • asymmetric_encrypt() and asymmetric_decrypt() use PKCS1 padding.

    • asymmetric_sign() and asymmetric_verify() use PKCS1 padding.

The `asymmetric_encrypt()` and `asymmetric_decrypt()` functions accept an extra optional parameter, padding. You can set it to `no`, `pkcs1`, or `oaep`. If you don’t specify this parameter, it defaults based on the `encryption_udf.legacy_padding_scheme` value. 

The padding schemes have the following limitations:

> | Padding Scheme    | Details                                                                                                                                                        |

Reviewer: Padding can only be specified for `asymmetric_decrypt()` / `asymmetric_encrypt()` when `algorithm` is `RSA`
There are limitations on the size of the message being encrypted for different RSA padding modes:
* for `no` the message length must be match the size of the RSA key in bytes (i.e. 128 bytes for 1024-bit RSA)
* for `pkcs1` message length must be no bigger than the size of the RSA key in bytes -11 (i.e. <=117 bytes for 1024-bit RSA)
* for `oaep` message length must be no bigger than the size of the RSA key in bytes -42 (i.e. <=86 bytes for 1024-bit RSA)

|-------------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------|
| `OAEP`           | With Optimal Asymmetric Encryption Padding (OAEP), the maximum message length is `<key_size_in_bytes> - 42 bytes`. This padding works exclusively with the RSA algorithm. |
| `no`             | The `no` padding scheme is strict. The message size must match the exact key size in bytes (e.g., 128 bytes for a 1024-bit RSA key). Other lengths generate an error. |

Similarly, `asymmetric_sign()` and `asymmetric_verify()` also have an optional padding parameter, which can be either `pkcs1` or `pkcs1_pss`. If not explicitly set, it follows the default based on `encryption_udf.legacy_padding_scheme`. You can only use the padding parameter with RSA algorithms.

> 

Reviewer: Probably, it's still makes sense to add a reference to the blog post
"Digital Signatures: Another Layer of Data Protection in Percona Server for MySQL"
https://www.percona.com/blog/digital-signatures-another-layer-of-data-protection-in-percona-server-for-mysql/
as an extended usage example.

The range for this variable is from 1,024 to 9,984. The default value is 9,984.

### encryption_udf.legacy_paddding_scheme

The variable enables or disables the legacy padding scheme for certain encryption operations.

| Option       | Description      |
|--------------|------------------|
| command-line | Yes              |
| scope        | Global           |
| data type    | Boolean |
| default      | OFF            |

This system variable is a BOOLEAN type and is set to `OFF` by default. 

This variable controls how the functions `asymmetric_encrypt()`, `asymmetric_decrypt()`, `asymmetric_sign()`, and `asymmetric_verify()` behave when you don’t explicitly set the padding parameter.

•	When encryption_udf.legacy_padding_scheme is OFF:

>     • asymmetric_encrypt() and asymmetric_decrypt() use OAEP padding.

Reviewer: ```suggestion
    • asymmetric_encrypt() and asymmetric_decrypt() use OAEP encryption padding.
```

>     • asymmetric_sign() and asymmetric_verify() use PKCS1_PSS padding.

Reviewer: ```suggestion
    • asymmetric_sign() and asymmetric_verify() use PKCS1_PSS signature padding.
```

•	When encryption_udf.legacy_padding_scheme is ON:

>     • asymmetric_encrypt() and asymmetric_decrypt() use PKCS1 padding.

Reviewer: ```suggestion
    • asymmetric_encrypt() and asymmetric_decrypt() use PKCS1 encryption padding.
```

>     • asymmetric_sign() and asymmetric_verify() use PKCS1 padding.

Reviewer: ```suggestion
    • asymmetric_sign() and asymmetric_verify() use PKCS1 signature padding.
```

The padding schemes have the following limitations:

| Padding Scheme    | Details                                                                                                                                                        |
|-------------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------|
| `oeap`           | The message you encrypt can be as long as your RSA key size in bytes - 42 bytes.|
| `no`             | The message length must exactly match your RSA key size in bytes. For example, if your key is 1024 bits (128 bytes), the message must also be 128 bytes. If it doesn’t match, it will cause an error. |
| `pkcs1` | Your message can be equal to or smaller than the RSA key size - 11 bytes. For instance, with a 1024-bit RSA key, your message can’t be longer than 117 bytes.|

> Similarly, `asymmetric_sign()` and `asymmetric_verify()` also have an optional padding parameter, which can be either `pkcs1` or `pkcs1_pss`. If not explicitly set, it follows the default based on [`encryption_udf.legacy_padding_scheme`](#encryption_udf.legacy_padding_scheme). You can only use the padding parameter with RSA algorithms.

Reviewer: ```suggestion
Similarly, `asymmetric_sign()` and `asymmetric_verify()` also have an optional `padding` parameter, which can be either `pkcs1` or `pkcs1_pss`. If not explicitly set, it follows the default based on [`encryption_udf.legacy_padding_scheme`](#encryption_udf.legacy_padding_scheme). You can only use the padding parameter with RSA algorithms.
```
